File: android/app/src/main/kotlin/cc/bsns/ssh/CommandHistory.kt
package cc.bsns.ssh

import android.content.Context
import org.json.JSONArray

/**
 * A local, on-device history of commands you've run — the privacy-respecting
 * counterpart to cloud "AI autocomplete": suggestions come from your own past
 * commands, nothing is uploaded. Stays on the device (deliberately NOT synced).
 */
class CommandHistory(context: Context) {
    private val prefs = context.getSharedPreferences("cmd_history", Context.MODE_PRIVATE)
    private val cap = 300

    /** Most-recent first. */
    fun all(): List<String> = try {
        val arr = JSONArray(prefs.getString("list", "[]"))
        (0 until arr.length()).map { arr.getString(it) }
    } catch (e: Exception) { emptyList() }

    /** Record a run command: move-to-front, de-duplicated, capped. */
    fun record(command: String) {
        val cmd = command.trim()
        if (cmd.isEmpty() || cmd.length > 1000) return
        val list = ArrayList(all())
        list.remove(cmd)
        list.add(0, cmd)
        while (list.size > cap) list.removeAt(list.size - 1)
        prefs.edit().putString("list", JSONArray(list).toString()).apply()
    }

    fun clear() = prefs.edit().remove("list").apply()

    /** History entries that extend `prefix` (prefix match, excluding an exact hit). */
    fun suggestions(prefix: String, limit: Int = 3): List<String> {
        val p = prefix.trimStart()
        if (p.length < 2) return emptyList()
        return all().asSequence()
            .filter { it.length > p.length && it.startsWith(p) }
            .distinct()
            .take(limit)
            .toList()
    }
}

File: app/Sources/Settings/CommandHistory.swift
import SwiftUI

/// A local, on-device history of commands you've run — the privacy-respecting
/// counterpart to cloud "AI autocomplete": entries come from your own past
/// commands, nothing is uploaded. Stays on the device (deliberately not synced).
/// Thread-safe: `TerminalSession.write` records from its I/O thread, the UI reads
/// snapshots on the main thread.
final class CommandHistory: @unchecked Sendable {
    static let shared = CommandHistory()

    private let lock = NSLock()
    private var entries: [String] = []
    private let cap = 300
    private let url: URL = {
        let dir = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("cmd_history.json")
    }()

    init() {
        if let data = try? Data(contentsOf: url),
           let decoded = try? JSONDecoder().decode([String].self, from: data) { entries = decoded }
    }

    /// Most-recent first.
    func all() -> [String] { lock.lock(); defer { lock.unlock() }; return entries }

    func record(_ command: String) {
        let cmd = command.trimmingCharacters(in: .whitespaces)
        guard !cmd.isEmpty, cmd.count <= 1000 else { return }
        lock.lock()
        entries.removeAll { $0 == cmd }
        entries.insert(cmd, at: 0)
        if entries.count > cap { entries.removeLast(entries.count - cap) }
        let snapshot = entries
        lock.unlock()
        try? JSONEncoder().encode(snapshot).write(to: url)
    }

    func clear() {
        lock.lock(); entries = []; lock.unlock()
        try? FileManager.default.removeItem(at: url)
    }
}

/// A picker over the local command history: tap to re-run, or clear.
struct CommandHistoryPicker: View {
    let onPick: (String) -> Void
    @Environment(\.dismiss) private var dismiss
    @State private var items: [String] = []
    @State private var query = ""

    private var filtered: [String] {
        query.isEmpty ? items : items.filter { $0.localizedCaseInsensitiveContains(query) }
    }

    var body: some View {
        NavigationStack {
            Group {
                if items.isEmpty {
                    ContentUnavailableView("No history yet", systemImage: "clock.arrow.circlepath",
                                           description: Text("Run a few commands and they'll show up here."))
                } else {
                    List(filtered, id: \.self) { cmd in
                        Button { onPick(cmd); dismiss() } label: {
                            Text(cmd).font(.system(.body, design: .monospaced)).lineLimit(1)
                        }
                    }
                    .searchable(text: $query)
                }
            }
            .navigationTitle("Command history")
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("Close") { dismiss() } }
                if !items.isEmpty {
                    ToolbarItem(placement: .destructiveAction) {
                        Button("Clear", role: .destructive) { CommandHistory.shared.clear(); items = [] }
                    }
                }
            }
            .onAppear { items = CommandHistory.shared.all() }
        }
    }
}
